File: Source/Bifrost.JavaScript/interaction/Operation.js
Bifrost.namespace("Bifrost.interaction", {
    Operation: Bifrost.Type.extend(function (region, context) {
        /// <summary>Defines an operation that be performed</summary>
        var self = this;

        /// <field name="context" type="Bifrost.interaction.Operation">Context in which the operation exists in</field>
        this.context = context;

        /// <field name="identifier" type="Bifrost.Guid">Unique identifier for the operation instance<field>
        this.identifier = Bifrost.Guid.empty;

        /// <field name="region" type="Bifrost.views.Region">Region that the operation was created in</field>
        this.region = region;

        /// <field name="canPerform" type="observable">Set to true if the operation can be performed, false if not</field>
        this.canPerform = ko.observable(true);
        
        this.perform = function () {
            /// <summary>Function that gets called when an operation gets performed</summary>
            /// <returns>State change, if any - typically helpful when undoing</returns>
            return {};
        };

        this.undo = function (state) {
            /// <summary>Function that gets called when an operation gets undoed</summary>
            /// <param name="state" type="object">State generated when the operation was performed</param>
        };
    })
});

File: Source/Bifrost.Client/Interaction/Operation.cs
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) 2008-2017 Dolittle. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
namespace Bifrost.Interaction
{
    /// <summary>
    /// Represents an operation in the system
    /// </summary>
    public class Operation
    {
        /// <summary>
        /// Perform the operation
        /// </summary>
        /// <param name="context"><see cref="OperationContext"/> in which the operation is performed in</param>
        /// <returns>Any state as a result of performing the operation</returns>
        public virtual OperationState Perform(OperationContext context)
        {
            return null;
        }

        /// <summary>
        /// Check if the operation can be performed
        /// </summary>
        /// <param name="context"><see cref="OperationContext"/> in which the operation is performed in</param>
        /// <returns>True if it can be performed, false if not</returns>
        public virtual bool CanPerform(OperationContext context)
        {
            return false;
        }


        /// <summary>
        /// Undo the given operation based on the state coming out of the perform
        /// </summary>
        /// <param name="context"><see cref="OperationContext"/> in which the operation was performed in</param>
        /// <param name="state">State as a result from when it was performed</param>
        public virtual void Undo(OperationContext context, OperationState state)
        {
        }

        /// <summary>
        /// Check if the operation can be undoed
        /// </summary>
        /// <param name="context"><see cref="OperationContext"/> in which the operation was performed in</param>
        /// <returns>True if it can be undoed, false if not</returns>
        public virtual bool CanUndo(OperationContext context)
        {
            return false;
        }
    }
}
